### DISCUSSION TITLE
Wasm bindgen interface?

### DISCUSSION THREAD

**Discussion by abhay-agarwal** [electric-sql/pglite#75] [General]
We've had this discussion in other database-in-the-browser projects (see https://github.com/rhashimoto/wa-sqlite/discussions/154), but I am wondering if anyone has any thoughts about having the database wasm be directly interoperable with a Wasm_bindgen-based Rust codebase.

There are a few ways that this might be done:
- PGLite is somehow built with some headers/linkers intact so that a wasm_bindgen compiler could link against it
- Wasm_bindgen loads the PGLite .wasm binary as a "wasm in wasm" strategy where the binary is loaded in and an interface is written in rust that communicates directly with that webassembly module. I don't personally know the performance implications of this strategy
- Emscipten is used end-to-end and rust users have to compile against a `wasm32-unknown-emscripten` target for their own codebases, which is really painful

Anyone have any thoughts on how one might go about this?

**Comment by samwillis** [comment]
Hey, one possibility is connecting to PGlite at the Postgres wire protocol layer, that way you will receive the raw byte arrays that can then be parsed directly to Rust types. That would remove the overhead of going Postgres types -> JS types -> Rust types, JS code would only be for connecting the parts and the type conversion becomes Postgres types -> Rust types.

It's also likely you would want to run PGlite in a separate worker to your main Rust code so you're not blocking it, with that you would be operating at the JS layer anyway.

We have an `execProtocol` method that enables this workflow. We haven't documented it yet as its design may change a little and I'm looking for feedback on how it could be used.

I've not looked, but I'm sure there will be a rust crate that contain the parts needed for constructing and parsing Postgres wire protocol messages.

**Reply by abhay-agarwal** [reply]
Amazing, I had no idea about the wire protocol. If that can speak in ArrayBuffer with another web worker then you’re off to the races. With Comlink + Wasm Bindgen to round out the devex. The main need in that case would be a postgres api-only crate or ORM that speaks the wire protocol and can take a hook that will postmessage everything to the DB. 

Cool stuff, will keep an eye out.
